spring-cloud-kubernetes-client-discovery/src/main/java/org/springframework/cloud/kubernetes/client/discovery/KubernetesInformerDiscoveryClient.java

```diff
@@ -145,39 +159,61 @@ private Stream<ServiceInstance> getServiceInstanceDetails(V1Service service, Str
-					.ofNullable(service.getMetadata().getLabels().get(PRIMARY_PORT_NAME_LABEL_KEY));
+				.ofNullable(service.getMetadata().getLabels().get(PRIMARY_PORT_NAME_LABEL_KEY));
 		}
 		final String primaryPortName = discoveredPrimaryPortName.orElse(this.properties.primaryPortName());
 
 		final boolean secured = isSecured(service);
 
-		return ep.getSubsets().stream().filter(subset -> subset.getPorts() != null && subset.getPorts().size() > 0) // safeguard
-				.flatMap(subset -> {
-					Map<String, String> metadata = new HashMap<>(svcMetadata);
-					List<CoreV1EndpointPort> endpointPorts = subset.getPorts();
-					if (this.properties.metadata() != null && this.properties.metadata().addPorts()) {
-						endpointPorts.forEach(
-								p -> metadata.put(StringUtils.hasText(p.getName()) ? p.getName() : UNSET_PORT_NAME,
-										Integer.toString(p.getPort())));
-					}
-					List<V1EndpointAddress> addresses = subset.getAddresses();
-					if (addresses == null) {
-						addresses = new ArrayList<>();
-					}
-					if (this.properties.includeNotReadyAddresses()
-							&& !CollectionUtils.isEmpty(subset.getNotReadyAddresses())) {
-						addresses.addAll(subset.getNotReadyAddresses());
-					}
-
-					final int port = findEndpointPort(endpointPorts, primaryPortName, serviceId);
-					return addresses.stream()
-							.map(addr -> new DefaultKubernetesServiceInstance(
-									addr.getTargetRef() != null ? addr.getTargetRef().getUid() : "", serviceId,
-									addr.getIp(), port, metadata, secured, service.getMetadata().getNamespace(),
-									// TODO find out how to get cluster name possibly from
-									// KubeConfig
-									null));
-				});
+		Map<String, V1ObjectMeta> finalPodMap = podMap;
+		return ep.getSubsets().stream()
+			.filter(subset -> subset.getPorts() != null && subset.getPorts().size() > 0) // safeguard
+			.flatMap(subset -> {
+				Map<String, String> metadata = new HashMap<>(svcMetadata);
+				List<CoreV1EndpointPort> endpointPorts = subset.getPorts();
+				if (this.properties.metadata() != null && this.properties.metadata().addPorts()) {
+					endpointPorts.forEach(
+						p -> metadata.put(StringUtils.hasText(p.getName()) ? p.getName() : UNSET_PORT_NAME,
+							Integer.toString(p.getPort())));
+				}
+				List<V1EndpointAddress> addresses = subset.getAddresses();
+				if (addresses == null) {
+					addresses = new ArrayList<>();
+				}
+				if (this.properties.includeNotReadyAddresses()
+					&& !CollectionUtils.isEmpty(subset.getNotReadyAddresses())) {
+					addresses.addAll(subset.getNotReadyAddresses());
+				}
+
+				final int port = findEndpointPort(endpointPorts, primaryPortName, serviceId);
+				return addresses.stream()
+					.map(addr -> new DefaultKubernetesServiceInstance(
+						addr.getTargetRef() != null ? addr.getTargetRef().getUid() : "", serviceId,
+						addr.getIp(), port, properties.initPodMetaData() ? initPodMetaData(addr, metadata, finalPodMap) : metadata, secured, service.getMetadata()
```

Comment:
IDK that we need a separate property, we already have a property to enable adding annotations and labels

Reply:
The metadata obtained here is the metadata of the pod, which should belong to metadata by definition